minimint/src/net/peers.rs

```diff
@@ -193,6 +195,267 @@ where
     }
 }
 
+struct CommonPeerConnectionState<M> {
+    resend_queue: MessageQueue<M>,
+    incoming: Sender<M>,
+    outgoing: Receiver<M>,
+    peer: PeerId,
+    cfg: ConnectionConfig,
+    connect: SharedAnyConnector<PeerMessage<M>>,
+    incoming_connections: Receiver<AnyFramedTransport<PeerMessage<M>>>,
+    last_received: Option<MessageId>,
+}
+
+struct DisconnectedPeerConnectionState<M> {
+    common: CommonPeerConnectionState<M>,
+    reconnect_at: Instant,
+    failed_reconnect_counter: u64,
+}
+
+struct ConnectedPeerConnectionState<M> {
+    common: CommonPeerConnectionState<M>,
```

Comment:
what if we flip this
```rust
struct Connection {
     resend_queue: MessageQueue<M>,
     incoming: Sender<M>,
     ....
     state: ConnectionState,
}

enum ConnectionState {
    Disconnected {
        ....
    },
    Connected {
        ....
    },
}
```

Reply:
Good idea, will try that!

Reply:
Got a bit distracted but tried to pull the common state out, do you think the current version is better? If so I'll build on that, add some docs to the new interfaces etc. Otherwise I'll go with the previous version as a base.

Reply:
This looks better to me.